File: Common/BoardUtils/HexUtils.cs
using System;
using System.Collections.Generic;
using Common.Data;
using Common.GameLogic;

namespace Common.BoardUtils
{
    public static class HexUtils
    {
        public static IEnumerable<VectorHex> ToHexMap(this string hex)
        {
            var zs = hex.Split('|');
            for (var z = 0; z < zs.Length; z++)
            {
                var yItem = zs[z].ToCharArray();
                for (var x = 0; x < yItem.Length; x++)
                {
                    var xItem = int.Parse(yItem[x].ToString());
                    yield return new VectorHex()
                    {
                        X = x,
                        Z = z,
                        Item = xItem
                    };
                }
            }
        }
        public static int Distance(GridHexagon p1, GridHexagon p2)
        {
            var x1 = p1.X;
            var y1 = p1.Z;

            var x2 = p2.X;
            var y2 = p2.Z;

            var du = x2 - x1;
            var dv = (y2 + ((x2 / 2) | 0)) - (y1 + ((x1 / 2) | 0));
            if ((du >= 0 && dv >= 0) || (du < 0 && dv < 0))
                return Math.Max(Math.Abs(du), Math.Abs(dv));
            else
                return Math.Abs(du) + Math.Abs(dv);
        }

        public static Direction GetDirection(GridHexagon p1, GridHexagon p2)
        {
            // console.log('x1', p1.x, 'x2', p2.x, 'y1', p1.z, 'y2', p2.z);
            string upDown = null;
            string leftRight = null;


            if (p1.X % 2 == 0)
            {
                if (p1.Z == p2.Z)
                {
                    upDown = "up";
                }
                else if (p1.Z < p2.Z)
                {
                    upDown = "down";
                }
                else if (p1.Z > p2.Z)
                {
                    upDown = "up";
                }
            }
            else
            {
                if (p1.Z == p2.Z)
                {
                    upDown = "down";
                }
                else if (p1.Z < p2.Z)
                {
                    upDown = "down";
                }
                else if (p1.Z > p2.Z)
                {
                    upDown = "up";
                }
            }


            if (p1.X < p2.X)
            {
                leftRight = "right";
            }
            else if (p1.X > p2.X)
            {
                leftRight = "left";
            }
            else
            {
                leftRight = "neither";
            }
            switch (leftRight)
            {
                case "left":
                    switch (upDown)
                    {
                        case "up":
                            return Direction.TopLeft;
                        case "down":
                            return Direction.BottomLeft;
                    }
                    break;
                case "right":
                    switch (upDown)
                    {
                        case "up":
                            return Direction.TopRight;
                        case "down":
                            return Direction.BottomRight;
                    }
                    break;
                case "neither":
                    switch (upDown)
                    {
                        case "up":
                            return Direction.Top;
                        case "down":
                            return Direction.Bottom;
                    }
                    break;
            }
            return Direction.Top;
        }

    }
}

File: Web/components/game/hexUtils.ts
import {GridHexagon} from "./gridHexagon";
export class Node {
    parent: Node = null;
    x = 0;
    z = 0;
    item: GridHexagon = null;
    f = 0;
    g = 0;

    constructor(parent: Node, piece: GridHexagon) {
        this.parent = parent;
        // array index of this Node in the world linear array

        // the location coordinates of this Node
        this.x = piece.x;
        this.z = piece.z;
        this.item = piece;
        // the distanceFunction cost to get
        // TO this Node from the START
        this.f = 0;
        // the distanceFunction cost to get
        // from this Node to the GOAL
        this.g = 0
    }

    value() {
        return this.x + (this.z * 5000);
    }
}

export interface Vector3 {
    x?: number;
    z?: number;
}


export class HexUtils {

    static distance(p1: Vector3, p2: Vector3) {
        const x1 = p1.x;
        const y1 = p1.z;

        const x2 = p2.x;
        const y2 = p2.z;

        const du = x2 - x1;
        const dv = (y2 + ((x2 / 2) | 0)) - (y1 + ((x1 / 2) | 0));
        if ((du >= 0 && dv >= 0) || (du < 0 && dv < 0))
            return Math.max(Math.abs(du), Math.abs(dv));
        else
            return Math.abs(du) + Math.abs(dv);
    }

    static orderBy<T>(list: T[], callback: (item: T) => number) {
        const itms = [];
        for (var i = 0; i < list.length; i++) {
            const obj = list[i];
            itms.push({item: obj, val: callback(obj)});
        }
        itms.sort((a, b) => (a.val - b.val));
        list = [];
        for (var i = 0; i < itms.length; i++) {
            const obj1 = itms[i];
            list.push(obj1.item);
        }
        return list;
    }


    static mathSign(f: number) {
        if (f < 0) return -1;
        else if (f > 0) return 1;
        return 0;
    }

    static getDirection(p1: Vector3, p2: Vector3): Direction {
        // console.log('x1', p1.x, 'x2', p2.x, 'y1', p1.z, 'y2', p2.z);
        var upDown: 'up' | 'down';
        var leftRight: 'left' | 'right' | 'neither';


        if (p1.x % 2 == 0) {
            if (p1.z === p2.z) {
                upDown = 'up';
            } else if (p1.z < p2.z) {
                upDown = 'down';
            } else if (p1.z > p2.z) {
                upDown = 'up';
            }
        } else {
            if (p1.z === p2.z) {
                upDown = 'down';
            } else if (p1.z < p2.z) {
                upDown = 'down';
            } else if (p1.z > p2.z) {
                upDown = 'up';
            }
        }


        if (p1.x < p2.x) {
            leftRight = "right";
        } else if (p1.x > p2.x) {
            leftRight = "left";
        } else {
            leftRight = "neither";
        }
        switch (leftRight) {
            case "left":
                switch (upDown) {
                    case "up":
                        return Direction.TopLeft;
                    case "down":
                        return Direction.BottomLeft;
                }
                break;
            case "right":
                switch (upDown) {
                    case "up":
                        return Direction.TopRight;
                    case "down":
                        return Direction.BottomRight;
                }
                break;
            case "neither":
                switch (upDown) {
                    case "up":
                        return Direction.Top;
                    case "down":
                        return Direction.Bottom;
                }
                break;
        }


    }
}

export enum Direction{
    Top = 0,
    TopRight = 1,
    BottomRight = 2,
    Bottom = 3,
    BottomLeft = 4,
    TopLeft = 5,
}